[SPARK-12091] [PYSPARK] Deprecate the JAVA-specific deserialized storage levels

gatorsmile opened the pull request:
The current default storage level of Python persist API is MEMORY_ONLY_SER. This is different from the default level MEMORY_ONLY in the official document and RDD APIs. 

@davies Is this inconsistency intentional? Thanks!

Updates: Since the data is always serialized on the Python side, the storage levels of JAVA-specific deserialization are not removed, such as MEMORY_ONLY. 

Updates: Based on the reviewers' feedback. In Python, stored objects will always be serialized with the [Pickle](https://docs.python.org/2/library/pickle.html) library, so it does not matter whether you choose a serialized level. The available storage levels in Python include `MEMORY_ONLY`, `MEMORY_ONLY_2`, `MEMORY_AND_DISK`, `MEMORY_AND_DISK_2`, `DISK_ONLY`, `DISK_ONLY_2` and `OFF_HEAP`.

gatorsmile committed:
Changed the default storage level of persist

gatorsmile committed:
Merge remote-tracking branch 'upstream/master' into persistStorageLevel

davies commented:
This is on purpose, see https://issues.apache.org/jira/browse/SPARK-2014 cc @mateiz

SparkQA commented:
**[Test build #47041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47041/consoleFull)** for PR 10092 at commit [`d52deb5`](https://github.com/apache/spark/commit/d52deb5b7eaf794d94f6ce87ce61b3410efbcc8f).

SparkQA commented:
**[Test build #47041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47041/consoleFull)** for PR 10092 at commit [`d52deb5`](https://github.com/apache/spark/commit/d52deb5b7eaf794d94f6ce87ce61b3410efbcc8f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47041/
Test PASSed.

gatorsmile commented:
@davies Thank you for showing me the original JIRA by @mateiz . It sounds like it does not make sense to keep data as deserialized Java objects since data is serialized on the Python side. Is my understanding correct?

I am wondering if we should automatically convert `MEMORY_ONLY` to `MEMORY_ONLY_SER` and `MEMORY_AND_DISK` to `MEMORY_AND_DISK_SER`, if users are choosing these two options? So far, the official document does not mention these issues. Users might not realize this when making a decision.  

Thank you!

mateiz commented:
It might be nice to only expose a smaller # of storage levels in Python, i.e. call them memory_only and memory_and_disk, but always use the serialized ones underneath.

gatorsmile commented:
@mateiz  Thank you for your answer! Will try to do it soon.

gatorsmile committed:
Removal of the JAVA-specific deserialized storage levels

SparkQA commented:
**[Test build #47116 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47116/consoleFull)** for PR 10092 at commit [`a6b7dd9`](https://github.com/apache/spark/commit/a6b7dd95722d80fd68a009fbcf331a586a84fb1a).

gatorsmile commented:
- Removed all the constants whose `deserialized` values are true. 
- Update the comments of StorageLevel
- Change the default storage levels of Kinesis level from `MEMORY_AND_DISK_2` to `MEMORY_AND_DISK_SER_2`.

Please verify if my changes are OK. @mateiz @davies Thank you very much!

gatorsmile commented:
Just saw the comments and will change the names soon. Thanks!

SparkQA commented:
**[Test build #47116 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47116/consoleFull)** for PR 10092 at commit [`a6b7dd9`](https://github.com/apache/spark/commit/a6b7dd95722d80fd68a009fbcf331a586a84fb1a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47116/
Test PASSed.

gatorsmile committed:
Renaming MEMORY_ONLY_SER to MEMORY_ONLY
Renaming MEMORY_ONLY_SER_2 to MEMORY_ONLY_2
Renaming MEMORY_AND_DISK_SER to MEMORY_AND_DISK
Renaming MEMORY_AND_DISK_SER_2 to MEMORY_AND_DISK_2

gatorsmile commented:
Based on the comments of @mateiz , the extra changes are made:
- Renaming MEMORY_ONLY_SER to MEMORY_ONLY
- Renaming MEMORY_ONLY_SER_2 to MEMORY_ONLY_2
- Renaming MEMORY_AND_DISK_SER to MEMORY_AND_DISK
- Renaming MEMORY_AND_DISK_SER_2 to MEMORY_AND_DISK_2

Thanks!

SparkQA commented:
**[Test build #47119 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47119/consoleFull)** for PR 10092 at commit [`0e074b6`](https://github.com/apache/spark/commit/0e074b6b7bde1705f04a360273b2915cdc1f383c).

davies commented on python/pyspark/storagelevel.py:
```diff
@@ -49,12 +51,8 @@ def __str__(self):
 StorageLevel.DISK_ONLY_2 = StorageLevel(True, False, False, False, 2)
-StorageLevel.MEMORY_ONLY = StorageLevel(False, True, False, True)
-StorageLevel.MEMORY_ONLY_2 = StorageLevel(False, True, False, True, 2)
-StorageLevel.MEMORY_ONLY_SER = StorageLevel(False, True, False, False)
```
Removing these will break backward compatibility, I'd like to deprecate them, explain the difference between Python and Java (say records will always serialized in Python)

gatorsmile replied:
Agree! Just updated the codes with the deprecated notes. Trying to follow the existing PySpark style. Please check if they are good. : )

Not sure if this will be merged to 1.6. The note is still using 1.6. Thank you!

davies replied:
It's too late for 1.6, and this change (API change) is good for 2.0, sounds good?

gatorsmile replied:
Sure. Just changed it. : )

SparkQA commented:
**[Test build #47119 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47119/consoleFull)** for PR 10092 at commit [`0e074b6`](https://github.com/apache/spark/commit/0e074b6b7bde1705f04a360273b2915cdc1f383c).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47119/
Test PASSed.

gatorsmile committed:
Add deprecated notes for the four deprecated storage levels.

gatorsmile committed:
Add deprecated notes for the four deprecated storage levels.

SparkQA commented:
**[Test build #47125 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47125/consoleFull)** for PR 10092 at commit [`014a3a8`](https://github.com/apache/spark/commit/014a3a8f31958bf1337a0c8df293fe15ac54cd9f).

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47124/
Test FAILed.

gatorsmile committed:
change the deprecated version from 1.6 to 2.0

SparkQA commented:
**[Test build #47129 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47129/consoleFull)** for PR 10092 at commit [`fef7ada`](https://github.com/apache/spark/commit/fef7ada1ec1d56c61de39c2268190bff6f6b0f42).

SparkQA commented:
**[Test build #47125 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47125/consoleFull)** for PR 10092 at commit [`014a3a8`](https://github.com/apache/spark/commit/014a3a8f31958bf1337a0c8df293fe15ac54cd9f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47125/
Test PASSed.

SparkQA commented:
**[Test build #47129 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/47129/consoleFull)** for PR 10092 at commit [`fef7ada`](https://github.com/apache/spark/commit/fef7ada1ec1d56c61de39c2268190bff6f6b0f42).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/47129/
Test PASSed.

gatorsmile commented:
Hi, @davies Will this be merged? or need more updates? Thanks! : )

davies commented:
@gatorsmile These changes looks good to me, could also update the docs/ (configuration and programming guide) to say that the storage level of Python RDD is different than Java/Scala ones?

gatorsmile commented:
Sure, will do it. Thanks!

gatorsmile committed:
update documents.

SparkQA commented:
**[Test build #48041 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48041/consoleFull)** for PR 10092 at commit [`3864f87`](https://github.com/apache/spark/commit/3864f873f1b9bc56a9cf8da46cb90e97b89bd291).

SparkQA commented:
**[Test build #48041 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/48041/consoleFull)** for PR 10092 at commit [`3864f87`](https://github.com/apache/spark/commit/3864f873f1b9bc56a9cf8da46cb90e97b89bd291).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/48041/
Test PASSed.

davies commented:
@gatorsmile LGTM, merging into master, thanks!
